### PR TITLE
Adopt YARP interfaces and implementations for CAN messages and bus drivers

### LINKED ISSUE #122: Replace current ICanBusHico interface with yarp::dev::ICanBus
Consider migrating [ICanBusHico](https://github.com/roboticslab-uc3m/yarp-devices/blob/cce12ef/libraries/BodyYarp/ICanBusHico.h) to [yarp::dev::ICanBus](http://www.yarp.it/classyarp_1_1dev_1_1ICanBus.html). This process would involve implementing new methods in the `CanBusHico` device and updating the type of current CAN interface handles (`canDevicePtr`).

### PR DESCRIPTION
**Pull request by PeterBowman**
Resolves #122. Main goals and improvements:

* Replace the `roboticslab::ICanBusHico` interface with [`yarp::dev::ICanBus`](http://www.yarp.it/classyarp_1_1dev_1_1ICanBus.html).
* Wrap CAN messages in [`yarp::dev::CanMessage`](http://www.yarp.it/classyarp_1_1dev_1_1CanMessage.html).
* New feature: [CAN acceptance filtering capability](http://www.cse.dmu.ac.uk/~eg/tele/CanbusIDandMask.html) (disabled by default) -> [**policy**](https://github.com/roboticslab-uc3m/yarp-devices/pull/158#issuecomment-404898303).
* Support for message buffers.
* TX/RX timeout values can be configured via YARP options (<=0 means *no timeout*) -> [**policy**](https://github.com/roboticslab-uc3m/yarp-devices/pull/158#issuecomment-405530212).
* Enable non-blocking read/write calls on demand.
* **[breaking change]** Now parses *real* bitrate values (e.g. `1000000`) instead of identifiers (e.g. `8`).

---

[2017/12/23] Needs testing, **do not merge**.
[2018/07/19] Tested on TEO.

### CODE REVIEW

**Comment by jgvictores** [comment]
This is great news and in fact something I wanted to do a long time ago... Thank you very much!!
(applying WIP label :-) )

**Comment by PeterBowman** [comment]
Much in the spirit of https://github.com/roboticslab-uc3m/yarp-devices/issues/74#issuecomment-302581336, I decided to pass the handle for `yarp::dev::ICanBufferFactory` to each CAN node device not via interface method (e.g. [roboticslab::ICanBusSharer::setCanBusPtr](https://github.com/roboticslab-uc3m/yarp-devices/blob/c05ad09ba26684d15af377b040941a17365276b5/libraries/YarpPlugins/ICanBusSharer.h)), but through YARP config properties. Both setter and getter are illustrated in the following snippets:

https://github.com/roboticslab-uc3m/yarp-devices/blob/c05ad09ba26684d15af377b040941a17365276b5/libraries/YarpPlugins/CanBusControlboard/DeviceDriverImpl.cpp#L91-L92

https://github.com/roboticslab-uc3m/yarp-devices/blob/c05ad09ba26684d15af377b040941a17365276b5/libraries/YarpPlugins/LacqueyFetch/DeviceDriverImpl.cpp#L21-L22

I'd probably need to add checks in node downstreams to assert presence and blob length for this property, much like it's done in [TechnosoftIpos/DeviceDriverImpl.cpp](https://github.com/roboticslab-uc3m/yarp-devices/blob/c05ad09ba26684d15af377b040941a17365276b5/libraries/YarpPlugins/TechnosoftIpos/DeviceDriverImpl.cpp) (e.g. `tr` and the like).

The major caveat, if I can call it so, is that it looks ugly and prone to errors if one simply forgets to add this single property upon device configuration. See:

https://github.com/roboticslab-uc3m/yarp-devices/blob/c05ad09ba26684d15af377b040941a17365276b5/example/cpp/exampleTechnosoftIpos/exampleTechnosoftIpos.cpp#L74-L86

Of course, the alternative is equally prone to omissions, although a bit closer to that C++ object-oriented feeling:

https://github.com/roboticslab-uc3m/yarp-devices/blob/c05ad09ba26684d15af377b040941a17365276b5/example/cpp/exampleTechnosoftIpos/exampleTechnosoftIpos.cpp#L141-L142

**Edit (18/07/2018):** not quite correct, see https://github.com/roboticslab-uc3m/yarp-devices/pull/158/commits/82c9067ac2322998d04b4c2f22593a1bdcbf91f1.

**Comment by jgvictores** [comment]
I like the idea of passing the pointer via blob. The additional setter methods were just hacks before I had experimented doing so on the openrave-yarp-plugins repo. :-)

**Comment by PeterBowman** [comment]
~~Libraries still get installed to `/usr/local/lib/i486-linux-gnu` despite roboticslab-uc3m/QA#39.~~

My bad, had to delete the cache and configure again.

File: libraries/YarpPlugins/CanBusHico/CanBusHico.cpp
Comment (PeterBowman) [inline libraries/YarpPlugins/CanBusHico/CanBusHico.cpp]: This is totally wrong. `CanBusHico::setDelay` must be repurposed to reflect one of three possible states:

* error (`ret < 0`)
* timeout (`ret == 0`), don't call `::read` afterwards
* success (`ret > 0`), tell callers to issue a `::read` operation

Also, I have to take a closer looks at every single call to `canRead` (perhaps `canWrite`, too) and ensure that it makes sense.

**Comment by PeterBowman** [comment]
[HiCO.CAN driver manual](https://github.com/roboticslab-uc3m/yarp-devices/blob/8be3065e6529536826e22aac41ff2e163a41d157/doc/hcanpci-linux-v1535en-edi1878.pdf), page 18:

> If the board internal filter table for the node (4 filters per CAN node) is full, this function will return with error `EBUSY`.

I'll need to implement range filters (or play with the filter mask).

**Comment by PeterBowman** [comment]
Disclaimer: CAN acceptance filters are **disabled by default** on this device (this is a new addition as stated in the PR description). The motivation behind this choice is that HiCO.CAN allows only 4 filters per CAN node (see [Linux manual](https://github.com/roboticslab-uc3m/yarp-devices/blob/ba82b176465f84b719a5b1e1fc2fc198af9d472f/doc/hcanpci-linux-v1535en-edi1878.pdf), page 18/28). In addition, two filter types are supported: masks and ranges. Masks are handled as expected, see https://github.com/roboticslab-uc3m/yarp-devices/issues/122#issuecomment-302578674. Be aware that the masking logic follows the dsPIC style (1a49082), whereas the manual mentions sja1000 at some point. Applying masks to arbitration ids do not take into account the function code in COB ids, for instance, the usual 0x580 or 0x600 numbers "prefixed" to any SDO operations carried out in the TechnosoftIpos devices (a full COB id would include the node identifier itself, e.g. 0x615 for node 15, and so on). In contrast, ranges work with lower and upper bounds, but do not take into account function codes and are therefore mostly useless in more complex CAN networks.

As a consequence, three modes have been introduced regarding acceptance filters:

* `disabled` - no support for acceptance filter, all messages are let through.
* `noRange` - only mask filters, limited to 4 nodes. Supports function codes.
* `maskAndRange` - you guess it. No function codes allowed.

The YARP config option that handles these modes is `canFilterConfiguration`.

**Comment by PeterBowman** [comment]
Sometimes, commands are correctly sent (`[ok]` acknowledge via RPC), but not handled correctly in the TechnosoftIpos. I observed this behaviour when working on a single joint: move, set reference speed, set reference acceleration (I didn't try more than that, though).

Console output on current develop and ASIBOT upon `set pos 4 5`:

```
...
[control mode lines for id 5 end here]
...
[debug] IPositionControl2Impl.cpp:20 positionMove(): (4, 5.000000)
[info] IPositionControl2RawImpl.cpp:17 positionMoveRaw(): (0,5.000000)
[success] IPositionControl2RawImpl.cpp:33 positionMoveRaw(): Sent "position target". 23 7a 60 0 8e 23 0 0. canId(5) via(600).
[info] ICanBusSharerImpl.cpp:310 interpretMessage(): Got SDO ack "position target" from driver. 60 7a 60 0 0 0 0 0. canId(5) via(580).
[success] IPositionControl2RawImpl.cpp:43 positionMoveRaw(): Sent "start position". 3f 0. canId(5) via(200).
[info] ICanBusSharerImpl.cpp:857 interpretMessage(): Got PDO1 that it is observed as ack "start position" from driver. 37 92. canId(5) via(180).
[info] ICanBusSharerImpl.cpp:903 interpretMessage(): Got PDO2 that it is observed as ack "start position" from driver. 37 92 1. canId(5) via(280).
[success] IPositionControl2RawImpl.cpp:55 positionMoveRaw(): Sent "reset position". f 0. canId(5) via(200).
```

All put nicely on consecutive lines. While issuing the same command on the HEAD of this PR, the robot does not move:

```
[success] IControlMode2RawImpl.cpp:183 getControlModeRaw1(): Sent modes of operation display. 40 61 60 0 0 0 0 0. canId(5) via(600).
[info] ICanBusSharerImpl.cpp:699 interpretMessage(): Got SDO ack "Detailed Error Register" from driver. 4b 2 20 0 0 0 0 0. canId(4) via(580).
[info] ICanBusSharerImpl.cpp:320 interpretMessage(): Got SDO "modes of operation display" from driver. 4f 61 60 0 1 0 0 0. canId(5) via(580).
[info] ICanBusSharerImpl.cpp:360 interpretMessage(): 	-Profile Position Mode. canId: 5.
[debug] IPositionControl2Impl.cpp:20 positionMove(): (4, 5.000000)
[info] IPositionControl2RawImpl.cpp:17 positionMoveRaw(): (0,5.000000)
[success] IPositionControl2RawImpl.cpp:33 positionMoveRaw(): Sent "position[warning] ICanBusSharerImpl.cpp:274 interpretMessage(): pt movement ended. canId: 5 (via 180).
[warning] ICanBusSharerImpl.cpp:274 interpretMessage(): pt movement ended. canId: 5 (via 280).
 target". 23 7a 60 0 8e 23 0 0. canId(5) via(600).
[success] IControlMode2RawImpl.cpp:215 getControlModeRaw2(): Sent manufacturer status query. 40 2 10 0 0 0 0 0. canId(5) via(600).
[info] ICanBusSharerImpl.cpp:556 interpretMessage(): Got "manufacturer status register" from driver. 43 2 10 0 37 86 1f 2. canId(5) via(580).
[info] ICanBusSharerImpl.cpp:560 interpretMessage(): 	-Ready to switch on. canId: 5.
[info] ICanBusSharerImpl.cpp:564 interpretMessage(): 	-Switched on. canId: 5.
[info] ICanBusSharerImpl.cpp:568 interpretMessage(): 	-Operation Enabled. canId: 5.
[info] ICanBusSharerImpl.cpp:576 interpretMessage(): 	-Motor supply voltage is present. canId: 5.
[info] ICanBusSharerImpl.cpp:600 interpretMessage(): 	-Remote: drive parameters may be modified via CAN and the drive will execute the command message. canId: 5.
[info] ICanBusSharerImpl.cpp:608 interpretMessage(): 	-Target reached. canId: 5.
[info] ICanBusSharerImpl.cpp:624 interpretMessage(): 	-No event set or the programmed event has not occurred yet. canId: 5.
[info] ICanBusSharerImpl.cpp:628 interpretMessage(): 	-Axis on. Power stage is enabled. Motor control is performed. canId: 5.
[info] ICanBusSharerImpl.cpp:637 interpretMessage(): 	*Drive/motor initialization performed. canId: 5.
[info] ICanBusSharerImpl.cpp:641 interpretMessage(): 	*Position trigger 1 reached. canId: 5.
[info] ICanBusSharerImpl.cpp:645 interpretMessage(): 	*Position trigger 2 reached. canId: 5.
[info] ICanBusSharerImpl.cpp:649 interpretMessage(): 	*Position trigger 3 reached. canId: 5.
[info] ICanBusSharerImpl.cpp:653 interpretMessage(): 	*Position trigger 4 reached. canId: 5.
[info] ICanBusSharerImpl.cpp:673 interpretMessage(): 	*Target command reached. canId: 5.
[info] ICanBusSharerImpl.cpp:556 interpretMessage(): Got "manufacturer status register" from driver. 43 2 10 0 37 86 1f 2. canId(5) via(580).
[info] ICanBusSharerImpl.cpp:560 interpretMessage(): 	-Ready to switch on. canId: 5.
[info] ICanBusSharerImpl.cpp:564 interpretMessage(): 	-Switched on. canId: 5.
[info] ICanBusSharerImpl.cpp:568 interpretMessage(): 	-Operation Enabled. canId: 5.
[info] ICanBusSharerImpl.cpp:576 interpretMessage(): 	-Motor supply voltage is present. canId: 5.
[info] ICanBusSharerImpl.cpp:600 interpretMessage(): 	-Remote: drive parameters may be modified via CAN and the drive will execute the command message. canId: 5.
[info] ICanBusSharerImpl.cpp:608 interpretMessage(): 	-Target reached. canId: 5.
[info] ICanBusSharerImpl.cpp:624 interpretMessage(): 	-No event set or the programmed event has not occurred yet. canId: 5.
[info] ICanBusSharerImpl.cpp:628 interpretMessage(): 	-Axis on. Power stage is enabled. Motor control is performed. canId: 5.
[info] ICanBusSharerImpl.cpp:637 interpretMessage(): 	*Drive/motor initialization performed. canId: 5.
[info] ICanBusSharerImpl.cpp:641 interpretMessage(): 	*Position trigger 1 reached. canId: 5.
[info] ICanBusSharerImpl.cpp:645 interpretMessage(): 	*Position trigger 2 reached. canId: 5.
[info] ICanBusSharerImpl.cpp:649 interpretMessage(): 	*Position trigger 3 reached. canId: 5.
[info] ICanBusSharerImpl.cpp:653 interpretMessage(): 	*Position trigger 4 reached. canId: 5.
[info] ICanBusSharerImpl.cpp:673 interpretMessage(): 	*Target command reached. canId: 5.
[success] IControlMode2RawImpl.cpp:231 getControlModeRaw3(): Sent Motion Error Register query. 40 0 20 0 0 0 0 0. canId(5) via(600).
[success] IPositionControl2RawImpl.cpp:43 positionMoveRaw(): Sent "start position". 3f 0. canId(5) via(200).
[success] IControlMode2RawImpl.cpp:247 getControlModeRaw4(): Sent Detailed Error Register query. 40 2 20 0 0 0 0 0. canId(5) via(600).
[info] IInteractionModeRawImpl.cpp:10 getInteractionModeRaw(): (0)
[info] IInteractionModeRawImpl.cpp:10 getInteractionModeRaw(): (0)
[info] IInteractionModeRawImpl.cpp:10 getInteractionModeRaw(): (0)
[info] IInteractionModeRawImpl.cpp:10 getInteractionModeRaw(): (0)
[info] IInteractionModeRawImpl.cpp:10 getInteractionModeRaw(): (0)
[success] IPositionControl2RawImpl.cpp:55 positionMoveRaw(): Sent "reset position". f 0. canId(5) via(200).
[info] ICanBusSharerImpl.cpp:862 interpretMessage(): Got PDO1 that it is observed when driver arrives to position target. 37 86. canId(5) via(180).
[info] ICanBusSharerImpl.cpp:908 interpretMessage(): Got PDO2 that it is observed when driver arrives to position target. 37 86 1. canId(5) via(280).
```

It becomes apparently obvious that the control mode request for this same node (id 5) and its response is mixed up with the actual position move command, possibly causing a disruption and conflicts at the `ICanBusSharer` level (or, even worse, in the iPOS itself). Whether this is a coincidence, I hadn't noticed it on develop, yet. Might be caused by how semaphores and thread locks are now handled within CanBusHico.

**Comment by PeterBowman** [comment]
Blocked on https://github.com/roboticslab-uc3m/asibot-main/issues/42.

**Comment by PeterBowman** [comment]
I was able to launch *oneCanBusOneWrapper* with no physical devices (i.e. *CanBusFake* as `canBusType` and *FakeJoint* as the node type in my .ini configuration file) thanks to https://github.com/roboticslab-uc3m/yarp-devices/commit/a19f9ba7f261af7b6a9fc86cc10629a574d903c6.

**Comment by PeterBowman** [comment]
I wonder if the `canOutputBuffer` instances introduced in each subdevice should be protected with a semaphore because of concurrent calls performed by ControlboardWrapper in its state thread. By contrast, there is only one `canInputBuffer` instance called from the main periodic thread (at CanBusControlboard).

**Comment by PeterBowman** [comment]
Keeping the future CanBusPeak device in mind, I'd rather drop the bitrate identifier (e.g. `8`) and accept the actual bitrate value (e.g. `1000000`) through the `canBitrate` RF option. Several .ini files across the &lt;robot&gt;-configuration-files repos should be updated thereafter.

**Comment by PeterBowman** [comment]
Idea: parse the `ids` option, if available, in the `open()` method. Thus, bulk filter additions will spare the need of performing several IO operations (the CanBusPeak device must reset upon applying a new filter).

**Comment by PeterBowman** [comment]
> Idea: parse the ids option, if available, in the open() method. Thus, bulk filter additions will spare the need of performing several IO operations (the CanBusPeak device must reset upon applying a new filter).

Well, this had been already done, at least in CanBusHico. Also, this device resets **all** filters on close. Commit https://github.com/roboticslab-uc3m/yarp-devices/pull/158/commits/46821732c86801643236e0e3eaa8ef78a2bd0714 made CanBusControlboard delete **individual** filters on close, which is done *before* the CAN device's own close method is called.

Idea: enforce a common policy for all (current and future) CAN devices regarding acceptance filters. Make them parse the `ids` option on init (may accept any other alias, too) and bulk load all filters. On close, bulk delete them. Callers (consumer apps) may use `yarp::dev::ICanBus::canIdAdd` and `yarp::dev::ICanBus::canIdDelete` on demand. It is not required that a filter should be set during the entire lifetime of the application. Some docs detail thay they are merely intended for quick debugging and filtering the terminal output for improved visualization.

I need to decide on what to do with CanBusControlboard.

**Comment by PeterBowman** [comment]
> I need to decide on what to do with CanBusControlboard.

As spoken with @jgvictores, this is our policy:

* CAN devices (e.g. CanBusHico, CanBusPeak) shall clear all active filters on `open()` and `close()`. The former can be avoided if consumers pass a `preserveFilters` option via RF mechanisms.
* Those devices will also parse an `ids` option from config and add filters, if available, for every CAN node id these values represent. Obviously, such operation must be done after clearing previously active filters (see previous point).
* The CanBusControlboard wrapper (and other consumer applications) may decide to configure all filters and delete them on `open()` and `close()` with `canIdAdd()` and `canIdDelete()`, respectively.
* On close, such devices or apps may issue `canIdDelete(0)`. Since an id `0` is illegal, it shall be interpreted by CAN devices as a "delete-all-filters" instruction.

**Comment by PeterBowman** [comment]
The aforementioned policy has been applied.

> Keeping the future CanBusPeak device in mind, I'd rather drop the bitrate identifier (e.g. `8`) and accept the actual bitrate value (e.g. `1000000`) through the `canBitrate` RF option. Several .ini files across the &lt;robot&gt;-configuration-files repos should be updated thereafter.

Done, added to description. This is a breaking change, that is, all affected .ini files in &lt;robot&gt;-configuration-files repos should be updated after the final merge.

**Comment by PeterBowman** [comment]
> I wonder if the `canOutputBuffer` instances introduced in each subdevice should be protected with a semaphore because of concurrent calls performed by ControlboardWrapper in its state thread. By contrast, there is only one `canInputBuffer` instance called from the main periodic thread (at CanBusControlboard).

Done.

**Comment by jgvictores** [comment]
In ye-oldie `CanBusHico` there was a [yarp::os::Semaphore](https://github.com/roboticslab-uc3m/yarp-devices/blob/ec902e4d92faafa0fd71618e2393a3a4c5037122/libraries/YarpPlugins/CanBusHico/CanBusHico.hpp#L68). Was never too sure if it was correctly implemented (I think it forces something like a half-duplex whereas CAN bus should be full-duplex), but maybe it's nice and cleaner to protect at that level?

**Comment by PeterBowman** [comment]
That semaphore avoids simultaneous reads/writes in the CAN device level, and it has been preserved. The new protections at the CAN node level (TechnosoftIpos, CuiAbsolute...) have been introduced because of the inherent YARP's ControlboardWrapper status thread, which would result in simultaneous writes both from the user side (e.g. a user requests the current pose via RPC) and the CB side (a background thread calls `getEncoders()` every few milliseconds). This PR adds a CAN message buffer which wasn't protected in any way until now. Hopefully, these new semaphores will help to solve the main blocker: https://github.com/roboticslab-uc3m/yarp-devices/pull/158#issuecomment-355359259.

**Comment by PeterBowman** [comment]
> Was never too sure if it was correctly implemented (I think it forces something like a half-duplex whereas CAN bus should be full-duplex) (...)

Although the previous behavior should be preserved in that respect, I have extended the scope of this semaphore so that the call to `::select()` (the mechanism that sets timeouts on a file descriptor) is also placed within the critical section. The nasty thing here is that write operations would have to wait for read timeouts to expire. On the other hand, read operations already had to wait for full-blocking writes (no timeout here) because of the same semaphore. I hope that being too careful does not impact performance (see https://github.com/roboticslab-uc3m/yarp-devices/issues/159).

**Comment by PeterBowman** [comment]
Both [`yarp::dev::ICanBus::canRead`](http://www.yarp.it/classyarp_1_1dev_1_1ICanBus.html#a31ceb6eb72e881bd7bff6c7310a2187b) and [`yarp::dev::ICanBus::canWrite`](http://www.yarp.it/classyarp_1_1dev_1_1ICanBus.html#a4c719ef9616012d3adcff9ca688b0ca8) accept a `wait` argument, e.g. `canRead(CanBuffer &msgs, unsigned int size, unsigned int *read, bool wait)`. This is meant to perform a blocking or non-blocking read or write call, respectively. I've come up with a not-so-elegant solution in this plugin (*CanBusHico*) using `fnctl` calls to change the blocking state of the hardware device. However, while working on *CanBusPeak* and the API its vendor provides, I noticed that there is no supported way to change such state during runtime (by contrast, there is a convenient parameter recognized on CAN open) without resorting to low level system calls, such as `fcntl`. Even if I implement a similar mechanism as in *CanBusHico*, I don't feel this is the right way - it seems like I'm trying to achieve more than the device is designed for. Or, perhaps, more than the `yarp::dev::ICanBus` methods are meant for.

Check the implementation of the [esdCan device](https://github.com/robotology/icub-main/blob/a5e94059e899c4fd6994d007dc7bdab9595a9111/src/libraries/icubmod/esdCan/EsdCan.cpp#L88-L114) at icub-main. I would infer that the sole purpose of accepting a `wait` parameter is to provide support for capable APIs. In that case, ESD offers a `canRead` function for blocking calls, and `canTake` for non-blocking (respectively, `canWrite` and `canSend`).

*CanBusPeak* is currently unable to automatically switch to blocking state on read/write if it wasn't properly initialized. As an alternative to implementing the switch-on-runtime mechanism from *CanBusHico*, I could introduce another configuration parameter - e.g. `canNonBlockingMode` - for telling the CAN device to initialize itself in blocking or non-blocking mode (defaulted to the former). Consumer apps (e.g. *CanBusControlboard*) would continue to issue `canRead`/`canWrite` with an implicit wait, if they want to, but CAN devices shall no longer care about having to forcefully switch to the requested mode if their API is not prepared for (cf. ESD in the previous example). In other words: don't try to conform to every piece of HW out there - resort to available stuff and configure it properly on init (via .ini file) -> embrace flexibility.

In a wider perspective, non-blocking read/write requests might be beneficial for achieving improved response times and overall device performance, see https://github.com/roboticslab-uc3m/yarp-devices/issues/159. At this point, I'm just not sure yet which way to choose.

**Comment by PeterBowman** [comment]
As spoken with @jgvictores (ASWJ), we have agreed on a new policy for CAN devices regarding (non-)blocking mode (not to confuse with the previous one regarding acceptance filters, https://github.com/roboticslab-uc3m/yarp-devices/pull/158#issuecomment-404898303):

* Such devices would open in blocking or non-blocking mode, depending on a `canBlockingMode` YARP option set to `1` (default) or `0`, respectively.
* If this mode does not match the expectations expressed in `canRead` or `canWrite`, that is, the `wait` parameter passed by callers (https://github.com/roboticslab-uc3m/yarp-devices/pull/158#issuecomment-405371706) is not equal to `canBlockingMode`, the device should return an error.
* The behavior defined in the previous point can be overriden. If `canAllowPermissive` is passed to the device via .ini option, it should go on with the requested read/write operation regardless of whether it is configured in blocking or non-blocking mode. Such option defaults to `0`.
* Already done: in blocking mode, read/write operations may wait the amount of milliseconds indicated by the `canRxTimeoutMs` and `canTxTimeoutMs` options, respectively. A value of zero or negative means that no timeout is appllied. Otherwise, stop blocking indefinitely unless data arrives to the receive buffer or the transmit buffer frees enough memory to hold more outgoing data.

**Comment by PeterBowman** [comment]
One word regarding the default parameter `wait` to `ICanBus::canRead`/`ICanBus::canWrite`: when writing our own implementation, care should be taken to use the exact same signature in the header file, e.g. in the `CanBusHico` class. That is:

```c++
bool roboticslab::CanBusHico::canRead(yarp::dev::CanBuffer & msgs, unsigned int size, unsigned int * read, bool wait = false);
```

Don't leave the `= false` out nor, even worse, change it to `true`. The former would implicitly use the parent's default upon calling these methods downstream:

```c++
iCanBus->canRead(canInputBuffer, msgs, &read); // assumes wait = false
```

The latter would definitely make this far more obscure: I don't remember which default is enforced, then I go to the `yarp::dev::ICanBus` docs and assume that it's `false`. Wrong! See https://stackoverflow.com/a/3533692.

**Comment by PeterBowman** [comment]
Regarding boolean-like options (`--canBlockingMode`, `--canAllowPermissive`): `0` or `1` would do if interpreted `.asBool()`, but `true` or `false` (via command line or .ini file) will work, too.

**Comment by PeterBowman** [comment]
Tested on TEO and working. We'll use this code on today's demos and leave it installed on the robot for future use. Let's speak about the implementation f2f.

**Review by jgvictores — APPROVED** [review]
LGTM!!

**Pull request merged**
